### PR TITLE
Fix TonY requests yarn config "yarn.io/gpu" on non-GPU clusters #500

### PR DESCRIPTION
**Pull request by oliverhu**
Fix TonY requests yarn config "yarn.io/gpu" on non-GPU clusters #500. Add a try catch block when querying GPU resource.

### CODE REVIEW

File: tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java [hunk 1]
```diff
@@ -1053,10 +1054,16 @@ public void onContainersAllocated(List<Container> containers) {
       for (Container container : containers) {
         // Need to explicitly remove container requests from remoteRequestsTable in AMRMClient, otherwise
         // resources get double-requested (YARN-1902)
+        int numGPU = 0;
+        try {
+          numGPU = (int) container.getResource().getResourceInformation(Constants.GPU_URI).getValue();
+        } catch (ResourceNotFoundException e) {
+          LOG.debug("No GPU resource configured for the cluster. ", e);
```
Comment (ashahab) [inline tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java:1061]: Is there a better way to check for no GPU resource in cluster? Throwing and caching exception seems extreme. 
See Item 69 from Josh Bloch's Effective Java: [Use exceptions only for exceptional conditions](https://ahdak.github.io/blog/effective-java-part-9/)

File: tony-core/src/main/java/com/linkedin/tony/util/Utils.java [hunk 2]
```diff
@@ -410,7 +410,9 @@ public static void addEnvironmentForResource(LocalResource resource, FileSystem
   public static AMRMClient.ContainerRequest setupContainerRequestForRM(JobContainerRequest request) {
     Priority priority = Priority.newInstance(request.getPriority());
     Resource capability = Resource.newInstance((int) request.getMemory(), request.getVCores());
-    Utils.setCapabilityGPU(capability, request.getGPU());
+    if (request.getGPU() > 0) {
+      Utils.setCapabilityGPU(capability, request.getGPU());
```
Comment (ashahab) [inline tony-core/src/main/java/com/linkedin/tony/util/Utils.java:414]: What happens with the prior behavior(Utils.setCapabilityGPU(capability, 0)?

Comment (plliao) [reply tony-core/src/main/java/com/linkedin/tony/util/Utils.java:414]: The function just returns with no harm.

File: tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java [hunk 3]
```diff
@@ -1074,10 +1076,15 @@ public void onContainersAllocated(List<Container> containers) {
...
+        int numGPU = 0;
+        if (ResourceUtils.getResourceTypeIndex().containsKey(Constants.GPU_URI)) {
+          numGPU = (int) container.getResource().getResourceInformation(Constants.GPU_URI).getValue();
+        }
+
         amRMClient.removeContainerRequest(Utils.setupContainerRequestForRM(new JobContainerRequest(
             "", 1, container.getResource().getMemorySize(),
             container.getResource().getVirtualCores(),
-            (int) container.getResource().getResourceInformation(Constants.GPU_URI).getValue(),
+            numGPU,
```
Comment (ashahab) [inline tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java:1087]: Is there a unit-test that ensures that this codepath works? We need one if there isn't one. We need to test both cases(if(true) and if(false)).

**Review by ashahab — CHANGES_REQUESTED** [review]

**Review by ashahab — APPROVED** [review]

**Comment by ashahab** [comment]
Approved, merge when ready.

**Pull request merged**